drivers/volume/portworx/portworx.go

```diff
@@ -1188,6 +1360,367 @@ func (p *portworx) ensureNodesDontMatchVersionPrefix(versionRegex *regexp.Regexp
+		podList = append(podList, pod)
+	}
+
+	failedPods, err := runCommandOnPods(podList, fmt.Sprintf("touch %s", killFile), execPodStepsHigh, false)
+
+	updateErr := updateRunningCommandPodListInSnap(snap, failedPods, taskID)
+	if updateErr != nil {
+		logrus.Warnf("failed to update list of pods with running command in snap due to: %v", updateErr)
+	}
+
+	return err
+}
+
+func hasSubset(set map[string]string, subset map[string]string) bool {
+	for k := range subset {
+		v, ok := set[k]
+		if !ok || v != subset[k] {
+			return false
+		}
+	}
+	return true
+}
+
+func runCommandOnPods(pods []v1.Pod, cmd string, numRetries int, failFast bool) ([]v1.Pod, error) {
+	var wg sync.WaitGroup
+	backOff := wait.Backoff{
+		Duration: defaultRetryInterval,
+		Factor:   defaultRetryFactor,
+		Steps:    numRetries,
+	}
+	errChannel := make(chan podErrorResponse)
+	finished := make(chan bool, 1)
+
+	for _, pod := range pods {
+		wg.Add(1)
+		go func(pod v1.Pod, errRespChan chan podErrorResponse) {
+			defer wg.Done()
+			err := wait.ExponentialBackoff(backOff, func() (bool, error) {
+				ns, name := pod.GetNamespace(), pod.GetName()
+				pod, err := k8s.Instance().GetPodByUID(pod.GetUID(), ns)
+				if err != nil {
+					if err == k8s.ErrPodsNotFound {
+						logrus.Infof("pod with uuid: %s in namespace: %s is no longer present", string(pod.GetUID()), ns)
+						return true, nil
+					}
+
+					logrus.Warnf("failed to get pod: [%s] %s due to: %v", ns, string(pod.GetUID()))
+					return false, nil
+				}
+
+				_, err = k8s.Instance().RunCommandInPod([]string{"sh", "-c", cmd}, name, "", ns)
+				if err != nil {
+					logrus.Warnf("failed to run command: %s on pod: [%s] %s due to: %v", cmd, ns, name, err)
+					return false, nil
+				}
+
+				logrus.Infof("command: %s succeeded on pod: [%s] %s", cmd, ns, name)
+				return true, nil
+			})
+			if err != nil {
+				errChannel <- podErrorResponse{
+					pod: pod,
+					err: err,
+				}
+			}
+		}(pod, errChannel)
+	}
+
+	// Put the wait group in a go routine.
+	// By putting the wait group in the go routine we ensure either all pass
+	// and we close the "finished" channel or we wait forever for the wait group
+	// to finish.
+	//
+	// Waiting forever is okay because of the blocking select below.
+	go func() {
+		wg.Wait()
+		close(finished)
+	}()
+
+	failed := make([]v1.Pod, 0)
+	select {
+	case <-finished:
+		if len(failed) > 0 {
+			err := fmt.Errorf("command: %s failed on pods: %s", cmd, podsToString(failed))
+			return failed, err
+		}
+
+		logrus.Infof("command: %s finished successfully on all pods", cmd)
+		return nil, nil
+	case errResp := <-errChannel:
+		failed = append(failed, errResp.pod) // TODO also accumulate atleast the last error
+		if failFast {
+			return failed, fmt.Errorf("command: %s failed in pod: [%s] %s due to: %s",
+				cmd, errResp.pod.GetNamespace(), errResp.pod.GetName(), errResp.err)
+		}
+	}
+
+	return nil, nil
+}
+
+func runBackgroundCommandOnPods(pods []v1.Pod, cmd, taskID string) error {
```

Comment:
Not specific to portworx, move to a common place

Reply:
moved all command ops to it's own package